monero-sys/src/lib.rs

```diff
@@ -0,0 +1,1511 @@
+        let_cxx_string!(password = password.unwrap_or(""));
+        let network_type = network_type.into();
+        let kdf_rounds = Self::DEFAULT_KDF_ROUNDS;
+
+        let wallet_pointer = unsafe {
+            self.inner.pinned().openWallet(
+                &path,
+                &password,
+                network_type,
+                kdf_rounds,
+                std::ptr::null_mut(),
+            )
+        };
+
+        if wallet_pointer.is_null() {
+            anyhow::bail!("Failed to open wallet: got null pointer")
+        }
+
+        let raw_wallet = RawWallet::new(wallet_pointer);
+
+        let wallet = FfiWallet::new(raw_wallet, background_sync, daemon)
+            .context("Failed to initialize wallet")?;
+
+        Ok(wallet)
+    }
+
+    /// Set the address of the remote node ("daemon").
+    fn set_daemon_address(&mut self, address: &str) {
+        tracing::debug!(%address, "Updating wallet manager's remote node");
+
+        let_cxx_string!(address = address);
+
+        self.inner.pinned().setDaemonAddress(&address);
+    }
+
+    /// Check if a wallet exists at the given path.
+    pub fn wallet_exists(&mut self, path: &str) -> bool {
+        let_cxx_string!(path = path);
+        self.inner.pinned().walletExists(&path)
+    }
+}
+
+impl RawWalletManager {
+    fn new(inner: *mut ffi::WalletManager) -> Self {
+        Self { inner }
+    }
+
+    /// Get a pinned reference to the inner (c++) wallet manager.
+    /// This is a convenience function necessary because
+    /// the ffi interface mostly takes a Pin<&mut T> but
+    /// we haven't figured out how to hold that in the struct.
+    pub fn pinned(&mut self) -> Pin<&mut ffi::WalletManager> {
+        unsafe {
+            Pin::new_unchecked(
+                self.inner
+                    .as_mut()
+                    .expect("wallet manager pointer not to be null"),
+            )
+        }
+    }
+}
+
+impl FfiWallet {
+    const MAIN_ACCOUNT_INDEX: u32 = 0;
+
+    /// Create and initialize new wallet from a raw C++ wallet pointer.
+    fn new(inner: RawWallet, background_sync: bool, daemon: Daemon) -> anyhow::Result<Self> {
+        if inner.inner.is_null() {
+            anyhow::bail!("Failed to create wallet: got null pointer");
+        }
+
+        let mut wallet = Self { inner };
+        wallet
+            .check_error()
+            .context("Something went wrong while creating the wallet (not null pointer, though)")?;
+
+        tracing::debug!(address=%wallet.main_address(), "Initializing wallet");
+
+        wallet
+            .init(&daemon.address, daemon.ssl)
+            .context("Failed to initialize wallet")?;
+
+        tracing::debug!("Initialized wallet, setting daemon address");
+
+        wallet.set_daemon_address(&daemon.address)?;
+
+        if background_sync {
+            tracing::debug!("Background sync enabled, starting refresh thread");
+
+            wallet.start_refresh();
+            wallet.refresh_async();
```

Comment:
I don't think we need to start a one-time sync process here. The background sync should be enough.